src/secret_sharing/array.rs

```diff
@@ -0,0 +1,245 @@
+use std::{
+    array,
+    fmt::Debug,
+    ops::{Add, AddAssign, Mul, Neg, Sub, SubAssign},
+};
+
+use generic_array::{ArrayLength, GenericArray};
+use typenum::U32;
+
+use crate::{
+    ff::{Field, Serializable},
+    helpers::Message,
+    secret_sharing::{SharedValue, SharedValueArray},
+};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StdArray<V, const N: usize>([V; N]);
+
+// Introducing the `Eq` bound here is a wart. For some reason AdditiveShare is `Eq`, but
+// `SharedValue` is not. Probably all of these should be Eq, or none of them.
+impl<V: SharedValue + Eq, const N: usize> SharedValueArray<V> for StdArray<V, N> {
```

Comment:
There are a bunch of related issues here, and I'm not sure `SharedValueArray` is the key element of all of them -- I thought the major design decision was between something like `AdditiveShare<[T; N]>` and `AdditiveShare<T, N>`.

The purpose of `SharedValueArray` is to serve as the trait bound for the `Array` associated type. The associated type `SharedValue::Array` can't be constrained as a `SharedValue`, since that creates a cycle when trying to determine if something is a `SharedValue`.

In the current version, the `Array` associated type is no longer on `SharedValue`, so I guess it might be possible to merge the `SharedValueArray` type with `SharedValue`. I'm not sure I want to do that though -- it seems nice to have a distinction between "I'm working with a value" and "I'm working with a vector of values". There's also an analogous set of traits `Field` and `FieldArray` -- in that case there is the added concern that the array is not a field.

There is a broader issue of whether we want the APIs used to implement protocols to make vectorization explicit (`SharedValue<V, N>` and `*Array` traits that need to be referenced in vectorized code), or implicit (`SharedValue<[V; N]>` and the regular operations on `SharedValue` and `Field` become vectorized). I believe that making it explicit is less error-prone. Any time you use the API vectorization, it's definitely vectorization across records (vertical) -- and you can apply this vectorization to anything. Vectorization across values within a record (horizontal) needs to be coded explicitly in the protocol, since it requires knowledge of the structure of the data.

Some of this is also related to the question of what the differences are between BA*, Gf2Array, and GfNBit. I am hoping to have a proposal for that soon.

Reply:
For the BA type, I was adding an Array trait (CustomArray, ArrayAccess). Would it be enough to distinguish a shared value of a single value from an array by checking whether the array trait has been implemented?

Reply:
This is a circuitous answer, but we might be able to merge `ArrayAccess` and `SharedValueArray` into a single trait, and then I would say that yes, whatever the merged trait is can distinguish arrays from single values.

A difference between `ArrayAccess` and `SharedValueArray` is that `SharedValueArray` inherits a bunch of the arithmetic operation traits. Right now I believe some arithmetic operations are invoked on BAs via `Field` -- if the BAs stop implementing `Field`, what trait do you think they should get the arithmetic operations from? (There are also some explicit `std::ops::Not` bounds in the PRF IPA protocols, but naming individual arithmetic operations everywhere we need them gets tedious.)

Reply:
Almost all arithmetic operations come from sharedvalue except multiplication (which currently comes from field) and std::ops::Not. I think it makes sense to bundle them all together in an existing trait. I am fine moving forward with SharedValueArray as a merge of the previous sharedvalue and arrayaccess/customarray. For the multiplication and maybe the not operation we might still need separate solutions since we might not want to implement them for all sharedvaluearrays. On the other side we could also optimize for bitarray since most of the operations are boolean arithmetics (so we could ask for sharedvaluearray to implement std::ops::Not).